File: Net.py
import numpy as np

class Net:
    smoothing = 256
    def __init__(self, ID, cells, px, py):
        self.ID    = ID
        self.cells = cells
        self.size  = len(cells)
        self.px    = px
        self.py    = py
  
    def locate_pins(self, x, y):
        pins = {'x': x[self.cells] + self.px[self.cells], 
                'y': y[self.cells] + self.py[self.cells]}
        return(pins)
    
    def WAWL(self, x, y):
        '''
        @brief         : Computation of weighted average smooth approximation of the net's half-perimeter wirelength
        @param x       : current placement x coordinates
        @param y       : current placement y coordinates
        @param cellIDs : IDs of cells in the given net
        '''
        pins = self.locate_pins(x, y)
        (Smaxx, dSmaxx) = self.smooth_extremum(pins['x'],  self.smoothing)
        (Smaxy, dSmaxy) = self.smooth_extremum(pins['y'],  self.smoothing)
        (Sminx, dSminx) = self.smooth_extremum(pins['x'], -self.smoothing)
        (Sminy, dSminy) = self.smooth_extremum(pins['y'], -self.smoothing)
        W = (Smaxx - Sminx) + (Smaxy - Sminy)
        dW = {'x': dSmaxx - dSminx, 
              'y': dSmaxy - dSminy}
        return(W, dW)
    
    def pin_variance(self, x, y):
        pins = self.locate_pins(x, y)
        V = np.var(pins['x']) + np.var(pins['y'])
        dV = {'x': (2/self.size)*(pins['x'] - np.mean(pins['x'])), 
              'y': (2/self.size)*(pins['y'] - np.mean(pins['y']))}
        return(V, dV)
    
    def HPWL(self, x, y):
        pins = self.locate_pins(x, y)
        hpwl = (pins['x'].max() - pins['x'].min()) + (pins['y'].max() - pins['y'].min())
        return(hpwl)
    
    @staticmethod
    def smooth_extremum(z, t):
        '''
        @brief normalized weighted average smooth approximation of maximum or minimum
        @param z : any vector (assumed z.shape = (len(z),))
        @param t : the smoothing parameter (t>0: maximum, t<0: minimum)
        '''
        z_norm = np.linalg.norm(z, 2)

        u = z/z_norm
        v = np.exp(t*u)
        w = v/v.sum()

        S = np.dot(w, u)
        q = np.multiply(w, 1 + t*(u - S))

        S_  = z_norm*S
        dS_ = q + (S - np.dot(q, u))*u
        return(S_, dS_)

File: Net.m
classdef Net < handle
  properties
    id;
    cells;
    net_size;
  end

  properties(Constant)
    smooth_param = 256
  end

  methods
    function [obj] = Net(id, cells)
      obj.id       = id;
      obj.cells    = cells;
      obj.net_size = length(cells);
    end

    function [pins] = locate_pins(obj, x, y)
      pins.x = x(obj.cells);
      pins.y = y(obj.cells);
    end

    function [W, dW] = WAWL(obj, x, y)
      %% ------------- WAWL()
      %  Computation of value and gradient of the weighted-average smooth 
      %  approximation of the net's half-perimeter wirelength. 
      pins = obj.locate_pins(x, y);

      [maxSx, maxdSx] = obj.smooth_extrema(pins.x,  obj.smooth_param);
      [maxSy, maxdSy] = obj.smooth_extrema(pins.y,  obj.smooth_param);

      [minSx, mindSx] = obj.smooth_extrema(pins.x, -obj.smooth_param);
      [minSy, mindSy] = obj.smooth_extrema(pins.y, -obj.smooth_param);

      W   = maxSx  - minSx  + maxSy  - minSy;
      dW.x = maxdSx - mindSx;
      dW.y = maxdSy - mindSy;
    end

    function [V, dV] = pin_spread(obj, x, y)
      %% ------------- pin_spread()
      %  Computation of the value and gradient for sum of the variances of 
      %  the net's x and y coordinates.  
      pins  = obj.locate_pins(x, y);
      V     = var(pins.x) + var(pins.y);
      dV.x  = (2/obj.net_size)*(pins.x - mean(pins.x));
      dV.y  = (2/obj.net_size)*(pins.y - mean(pins.y));
    end

  end
  
  methods(Static)
    function [S_, dS_] = smooth_extrema(z, t)
      %% --------------- smooth_extrema()
      % The smoothing paramater determines whether the minimum (t < 0) or
      % the maximum (t > 0) of w is approximated. Note that the argument z
      % is normalized, the approximate maximum is computed, and then
      % rescaled. This is to prevent overflow in v = exp(t*u).
      z_norm = norm(z);
      
      u = z/z_norm;
      v = exp(t*u); 
      w = v/sum(v);

      S = w'*u; % Smooth max of the unit vector u
      q = w.*(1 + t*(u - S));
      
      S_  = z_norm*S;  
      dS_ = q + (S - q'*u)*u;
    end
  end
end